contrib/check-ifd-firmware.py

```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: LGPL-2.1+
+#
+# pylint: disable=invalid-name,missing-docstring,too-few-public-methods
+
+from typing import Dict
+import os
+import glob
+import sys
+import subprocess
+import json
+from collections import defaultdict
+
+
+class TermColor:
+    HEADER = "\033[95m"
+    OKBLUE = "\033[94m"
+    OKCYAN = "\033[96m"
+    OKGREEN = "\033[92m"
+    WARNING = "\033[93m"
+    FAIL = "\033[91m"
+    ENDC = "\033[0m"
+    BOLD = "\033[1m"
+    UNDERLINE = "\033[4m"
+
+
+def _scan_file(fn: str) -> Dict[str, int]:
+    new_map: Dict[str, int] = defaultdict(int)
+    try:
+        print(f"loading {fn}…")
+        args = [
+            "./src/fwupdtool",
+            "firmware-parse",
+            fn,
+            "ifd-firmware",
+            "--json",
+            "--no-timestamp",
+        ]
+        p = subprocess.run(args, check=True, capture_output=True)
+    except subprocess.CalledProcessError as e:
+        print(f"{' '.join(args)}: {e}")
+    else:
+        for line in p.stdout.decode().split("\n"):
+            new_map["Lines"] += 1
+            if line.find("gtype=") == -1:
+                continue
+            sections = line.split('"')
+            if not sections[1].startswith("Fu"):
+                continue
+            new_map[sections[1]] += 1
+        for line in p.stderr.decode().split("\n"):
+            if not line:
+                continue
+            new_map["WarningLines"] += 1
+            print(line)
+    return new_map
+
+
+def _scan_dir(path: str) -> bool:
+    all_okay: bool = True
+    needs_save: bool = False
+
+    results: Dict[str, Dict[str, int]] = {}
+    for fn in glob.glob(f"{path}/*.bin"):
+        results[fn] = _scan_file(fn)
+
+    # go through each result
+    print(f"    {os.path.basename(sys.argv[0])}:")
+    for fn, new_map in results.items():
+        try:
+            with open(f"{fn}.json", "rb") as f:
+                old_map = json.loads(f.read().decode())
+        except FileNotFoundError:
+            old_map = {}
+        print(f"    {fn}")
+        for key in sorted(set(list(old_map.keys()) + list(new_map.keys()))):
+            cnt_old = old_map.get(key, 0)
+            cnt_new = new_map.get(key, 0)
+            color = TermColor.HEADER
+            if cnt_new > cnt_old:
+                color = TermColor.OKGREEN
+                needs_save = True
+            elif cnt_new < cnt_old:
+                color = TermColor.FAIL
+                if key != "WarningLines":
+                    all_okay = False
+            else:
+                continue
+            print(f"    {color}{key:36}{TermColor.ENDC}: {cnt_old} -> {cnt_new}")
+
+    # save new results if all better
+    if needs_save and all_okay:
+        for fn, new_map in results.items():
+            with open(f"{fn}.json", "wb") as f:
+                f.write(json.dumps(new_map, sort_keys=True, indent=4).encode())
+
+    return all_okay
+
+
+if __name__ == "__main__":
+    rc: int = 0
+    for _path in sys.argv[1:]:
```

Comment:
Mostly for future selfs I think it would be good to change this to argparse so there could be a nice --help to indicate what is expected to be used in it.

Reply:
Good idea, added.